rfc/001-composable-pipelines.md

```diff
@@ -0,0 +1,292 @@
+been sorted out and analysts interact with the data, it becomes clear that some
+part of the data needs to be cleaned. This matches ELT mindset of dumping
+everything first so that the data is there, and then tweak as you go.
+
+Example operations include, migrating data to a different schema, enriching
+telemetry (e.g., with GeoIP data), and fixing invalid entries.
+
+Mutation of data is non-trivial in that it requires exclusive access to the
+data. Otherwise data races may occur. Consequently, we need transactional
+interface (ACID) to support this operation.
+
+As long as VAST owns the underlying data (i.e., only VAST is allowed to make
+changes), VAST can already mutate of data at rest. [Spatial
+compaction][mutate-at-rest] uses a pre-defined disk quota as trigger to apply
+pipelines to a subset of to-be-transformed data. After VAST applies the
+pipeline, VAST optionally removes the original data in an atomic fashion.
+
+[mutate-at-rest]: https://vast.io/docs/use-vast/transform#modify-data-at-rest
+
+Modeled after the `compaction` plugin, we may consider exposing a mutable
+pipeline interface through a dedicate command. This command interacts with a
+node, and could be seen as a combination of `pull` and `push` that executes
+remotely. For example:
+
+```bash
+vast mutate --types=A,B,C 'where 10.0.0.0/8 | put orig.h resp.h'
+```
+
+This would apply the pipeline to filter and project in schema types `A`, `B`,
+and `C`.
+
+## Implementation
+
+We could implement the entire pipeline with Arrow IPC streams in the "interior"
+operators, with only `from` and `to` taking raw input as byte stream. But this
+may be too restrictive. Let's assume that input and output are dynamically
+typed. An operator must then define its input and output types, e.g., `from<In,
+Out>` takes as input `In` and produces instances of `Out`.
+
+### Matcher Example
+
+This can come in handy for working with the matcher plugin. Consider the use
+case of constructing a matcher from data. Consider this pipeline:
+
+```bash
+vast exec 'read csv |
+           matcher build --extract=net.src.ip' > ips.flatbuffer
+```
+
+Here, we have:
+
+- `read<Bytes, Arrow>`: reads `Bytes` and produces `Arrow` record batches
+- `matcher build<Arrow, Flatbuffer>`: consumes `Arrow` and create a matcher
+  backend.
+
+Let `Void` be a valid type where the operator as only a side effect. Then we can
+pre-load a matcher at a VAST node as follows:
+
+```bash
+vast exec 'from file:///ips.flatbuffer |
+           matcher load --name=ips vast:///1.2.3.4'
+```
+
+Or more UNIX'ish, assuming an implicit `from -`:
+
+```bash
+vast exec 'matcher load --name=ips' < ips.flatbuffer
+```
+
+Here, `matcher load<Flatbuffer, Void>` is the operator that updates the remote
+matcher state.
+
+Finally, the `matcher` sub-command `match` does the actual matching:
+
+```bash
+vast pull 'where #type == "suricata.flow" |
+           matcher match --on=:addr ips'
+```
```

Comment:
Good observation. I think the remote vs. local aspect is not yet very clear in the current proposal.

You could argue that everything should run locally by default. If you want to go remote, you need to use something like `vast pipeline run`. 

You could also argue that all execution is *remote* unless you "break" the pipeline, by using a separate `vast` process:

```bash
vast pull 'where EXPR | summarize f(a) f(b) group-by x y z' |
  vast exec 'print csv'
```

This would only do the rendering client-side.

I'm not sure if the `Remote`/`Local` direction *at the operator level* simplifies things. Even though it's explicit, I feels like too much complexity (at least now).

Reply:
"everything is local" doesn't allow you to reconstruct the current `import` command as pointed out by @dominiklohmann [here](https://github.com/tenzir/vast/pull/2511/files#r955712536), except if you give the `pull` operator a pipeline as argument. I'll resolve this conversation so we centralize this discussion there.